File: Source/Core/Selections/Selection.cs
using System;

namespace Quip {
  /// <summary>
  /// A inclusive, directed range of locations within a document.
  /// </summary>
  public class Selection {
    /// <summary>
    /// Initializes a new instance.
    /// </summary>
    /// <param name="origin">The origin.</param>
    /// <param name="extent">The extent.</param>
    public Selection (Location origin, Location extent) {
      Origin = origin;
      Extent = extent;
    }

    /// <summary>
    /// Gets the selection's origin.
    /// </summary>
    public Location Origin {
      get;
      set;
    }

    /// <summary>
    /// Gets the selection's extent.
    /// </summary>
    /// <value>The extent.</value>
    public Location Extent {
      get;
      set;
    }

    public Location LowerBound {
      get {
        return Origin < Extent ? Origin : Extent;
      }
    }

    public Location UpperBound {
      get {
        return Origin < Extent ? Extent : Origin;
      }
    }

    public int Height {
      get {
        return UpperBound.Row - LowerBound.Row + 1;
      }
    }
  }
}

File: Source/Selection.cpp
#include "Selection.hpp"

#include <algorithm>

namespace quip {
  Selection::Selection (const Location & origin, const Location & extent)
  : m_origin(origin)
  , m_extent(extent) {
  }
  
  const Location & Selection::origin () const {
    return m_origin;
  }
  
  const Location & Selection::extent () const {
    return m_extent;
  }

  void Selection::setOrigin (const Location & location) {
    m_origin = location;
  }
  
  void Selection::setExtent (const Location & location) {
    m_extent = location;
  }
  
  const Location & Selection::lowerBound () const {
    return std::min(origin(), extent());
  }
  
  const Location & Selection::upperBound () const {
    return std::max(origin(), extent());
  }
  
  std::uint64_t Selection::height () const {
    return upperBound().row() - lowerBound().row() + 1;
  }
  
  bool operator== (const Selection & left, const Selection & right) {
    return left.origin() == right.origin() && left.extent() == right.extent();
  }
  
  bool operator!= (const Selection & left, const Selection & right) {
    return !(left == right);
  }
}